engine/src/test/java/io/zeebe/engine/processor/workflow/timer/ConcurrentTimerEventTest.java

```diff
@@ -0,0 +1,174 @@
+  }
+
+  @Test
+  public void shouldTriggerCommandIfTimerIsTriggered() {
+    // when
+    ENGINE.writeRecords(
+        RecordToWrite.command()
+            .timer(TimerIntent.TRIGGER, timerCreated.getValue())
+            .key(timerCreated.getKey()),
+        RecordToWrite.command()
+            .timer(TimerIntent.TRIGGER, timerCreated.getValue())
+            .key(timerCreated.getKey()));
+
+    ENGINE.start();
+
+    // then
+    final var rejection =
+        RecordingExporter.timerRecords(TimerIntent.TRIGGER)
+            .onlyCommandRejections()
+            .withWorkflowInstanceKey(workflowInstanceKey)
+            .getFirst();
+
+    Assertions.assertThat(rejection).hasRejectionType(RejectionType.NOT_FOUND);
+  }
+
+  @Test
+  public void shouldRejectCancelCommandIfTimerIsTriggered() {
+    // when
+    ENGINE.writeRecords(
+        RecordToWrite.command()
+            .timer(TimerIntent.TRIGGER, timerCreated.getValue())
+            .key(timerCreated.getKey()),
+        RecordToWrite.command()
+            .timer(TimerIntent.CANCEL, timerCreated.getValue())
+            .key(timerCreated.getKey()));
+
+    ENGINE.start();
+
+    // then
+    final var rejection =
+        RecordingExporter.timerRecords(TimerIntent.CANCEL)
+            .onlyCommandRejections()
+            .withWorkflowInstanceKey(workflowInstanceKey)
+            .getFirst();
+
+    Assertions.assertThat(rejection).hasRejectionType(RejectionType.NOT_FOUND);
+  }
+
+  @Test
+  public void shouldRejectTriggerCommandIfElementInstanceIsLeft() {
+    // given
+    final var processActivated =
+        RecordingExporter.workflowInstanceRecords(WorkflowInstanceIntent.ELEMENT_ACTIVATED)
+            .withElementType(BpmnElementType.PROCESS)
+            .getFirst();
+
+    final var eventActivated =
+        RecordingExporter.workflowInstanceRecords(WorkflowInstanceIntent.ELEMENT_ACTIVATED)
+            .withElementType(BpmnElementType.INTERMEDIATE_CATCH_EVENT)
+            .getFirst();
+
+    // when
+    ENGINE.writeRecords(
+        RecordToWrite.command()
+            .workflowInstance(WorkflowInstanceIntent.CANCEL, processActivated.getValue())
+            .key(processActivated.getKey()),
+        RecordToWrite.event()
+            .workflowInstance(
+                WorkflowInstanceIntent.ELEMENT_TERMINATING, processActivated.getValue())
+            .key(processActivated.getKey())
+            .causedBy(0),
+        RecordToWrite.event()
+            .workflowInstance(WorkflowInstanceIntent.ELEMENT_TERMINATING, eventActivated.getValue())
+            .key(eventActivated.getKey())
+            .causedBy(1),
+        RecordToWrite.event()
+            .workflowInstance(WorkflowInstanceIntent.ELEMENT_TERMINATED, eventActivated.getValue())
+            .key(eventActivated.getKey())
+            .causedBy(2),
+        RecordToWrite.command()
+            .timer(TimerIntent.TRIGGER, timerCreated.getValue())
+            .key(timerCreated.getKey()));
```

Comment:
I'm not sure I like this test. It looks like the test binds too much to the implementation. Instead of writing commands and events, I feel it would be better to only write commands and wait for something to have occurred before writing another command. For example, send the cancel command for the process, that in turn terminates the process incl the timer intermediate catch event. We wait for that to happen and then send a trigger timer command.

Reply:
I see your point. This type of tests are more coupled to the implementation (i.e. the records on the stream).  

However, I see no other way to test this. We have no way to pause the stream processing and inject commands.

If I would write the test like this:
* wait until the timer event is activated
* write the cancel workflow instance command
* wait until the workflow instance is terminated
* write the trigger timer command

then it wouldn't test the behavior because the timer is canceled when the timer event is terminated. The command would be rejected but it is a different situation. 

Let's discuss it f2f if you want to challenge the approach or have more questions :)

Reply:
As discussed, this approach was chosen after pausing the stream processing became too complex in other tests.